cpp/tests/rolling/rolling_test.cpp

```diff
@@ -0,0 +1,723 @@
+                                                                  min_periods, op));
+    }
+
+    auto reference = create_reference_output(op, input, window, forward_window, min_periods);
+
+#if 0
+    std::cout << "input:\n";
+    cudf::test::print(input, std::cout, ", ");
+    std::cout << "\n";
+    std::cout << "output:\n";
+    cudf::test::print(*output, std::cout, ", ");
+    std::cout << "\n";
+    std::cout << "reference:\n";
+    cudf::test::print(reference, std::cout, ", ");
+    std::cout << "\n";
+    std::cout << "\n";
+#endif
+
+    cudf::test::expect_columns_equal(*output, *reference);
+  }
+
+  // helper function to test all aggregators
+  void run_test_col_agg(cudf::column_view const& input,
+                        const std::vector<size_type> &window,
+                        const std::vector<size_type> &forward_window,
+                        size_type min_periods)
+  {
+    // test all supported aggregators
+    run_test_col(input, window, forward_window, min_periods, rolling_operator::MIN);
+    run_test_col(input, window, forward_window, min_periods, rolling_operator::MAX);
+    run_test_col(input, window, forward_window, min_periods, rolling_operator::COUNT);
+    run_test_col(input, window, forward_window, min_periods, rolling_operator::MEAN);
+
+    if (!cudf::is_timestamp(input.type()))
+      run_test_col(input, window, forward_window, min_periods, rolling_operator::SUM);
+  }
+
+  private:
+
+  // use SFINAE to only instantiate for supported combinations
+
+  // specialization for COUNT
+  std::unique_ptr<cudf::column> 
+  create_count_reference_output(cudf::column_view const& input,
+                                std::vector<size_type> const& window_col,
+                                std::vector<size_type> const& forward_window_col,
+                                size_type min_periods)
+  {
+    size_type num_rows = input.size();
+    std::vector<cudf::size_type> ref_data(num_rows);
+    std::vector<bool> ref_valid(num_rows);
+
+    // input data and mask
+  
+    std::vector<bitmask_type> in_valid; 
+    //std::tie(in_col, in_valid) = cudf::test::to_host<T>(input); 
+    in_valid = cudf::test::bitmask_to_host(input);
+    bitmask_type* valid_mask = in_valid.data();
+    
+    for(size_type i = 0; i < num_rows; i++) {
+      size_type count = 0;
+      // load sizes
+      min_periods = std::max(min_periods, 1); // at least one observation is required
+
+      // compute bounds
+      auto window = window_col[i%window_col.size()];
+      auto forward_window = forward_window_col[i%forward_window_col.size()];
+      size_type start_index = std::max((size_type)0, i - window + 1);
+      size_type end_index = std::min(num_rows, i + forward_window + 1); // exclusive
+      
+      // aggregate
+      for (size_type j = start_index; j < end_index; j++) {
+        if (!input.nullable() || cudf::bit_is_set(valid_mask, j))
+          count++;
+      }
```

Comment:
Unclear to me why `!input.nullable()` is inside loop. If `j` does't impact `input` looks like the code should be something like:
```cpp
      size_type count = [&] () {
        if (!input.nullable())
          end_index - start_index;
        else
          for (size_type j = start_index; j < end_index; j++)
            if (cudf::bit_is_set(valid_mask, j))
              count++;
      } ();
```
Also, the `else` is pretty clearly a `count_if` but based on looking at `bitmask_type` being `uint8_t` and how `cudf::bit_is_set` works, I don't think we can use it.

Reply:
`bitmask_type` is `uint32_t`

https://github.com/rapidsai/cudf/blob/3d9da9e5da2e731d3d63a066a6d44fe3465cbb5a/cpp/include/cudf/types.hpp#L84

Reply:
sorry, yes that is what i meant

Reply:
I didn't want to diverge this code too far from the kernel that does the same thing on the GPU.  I like the first suggestion though, as it can work on both.

Reply:
BTW, we have exactly the function you want, called `cudf::count_set_bits(mask, start, end)`, but it expects a device pointer (runs a kernel). This is test code which runs on the CPU for verification.